### PR TITLE
moving modules to ts to make it easier to code review / maintain

### PR DESCRIPTION
**Pull request by basarat**
:rose:

### CODE REVIEW

**Comment by basarat** [comment]
Alright. Ready for review :rose:

**Comment by basarat** [comment]
Side note: How do you feel about me moving off of `gulp` to just npm scripts. Getting gulp to fail build on compile errors is harder than it needs to be https://github.com/ivogabe/gulp-typescript/issues/78#issuecomment-78536581 ? :rose:

**Comment by basarat** [comment]
> Getting gulp to fail build on compile errors is harder than it needs to be ivogabe/gulp-typescript#78 (comment) ?

Nevermind. Got it working

**Comment by basarat** [comment]
⚠️  Do not merge. Making sure that builds fail correctly :rose:

**Comment by basarat** [comment]
:memo: let me know if you want to still use `hello.js` / its watch and I'll bring it back with more docs :rose:

File: modules/fusebox-hot-reload/index.js [hunk 1]
```diff
@@ -1,46 +0,0 @@
...
-                if (FuseBox.mainFile) {
-                    FuseBox.import(FuseBox.mainFile)
-                }
-            }
-            if (data.type === "css" && __fsbx_css) {
-                __fsbx_css(data.path, data.content)
-            }
-        })
-        client.on("error", (erro) => {
-            console.log(error);
```
Comment (basarat) [inline modules/fusebox-hot-reload/index.js]: :bug: was a typo

File: gulpfile.js [hunk 2]
```diff
@@ -124,23 +137,23 @@ gulp.task('npm-publish', function(done) {
...
-gulp.task('dist', ['dist-typings', 'dist-commonjs', 'dist-loader', 'dist-modules'], function() {});
+gulp.task('dist', ['dist-main', 'dist-loader', 'dist-modules']);
 
 /**
  * For development workflow
  */
 gulp.task('watch', ['dist'], function() {
     watching = true;
 
-    gulp.watch(['dist-loader/**/*.ts'], () => {
```
Comment (basarat) [inline gulpfile.js]: 🐛 fixed to point to new loader location :rose:

File: gulpfile.js [hunk 3]
```diff
@@ -73,18 +83,21 @@ gulp.task('minify-loader', function() {
         .pipe(rename('fusebox.min.js')).pipe(gulp.dest('modules/fuse-box-loader-api'))
 });
 
+/**
+ * Main building
+ */
 gulp.task('dist-typings', () => {
-    let result = gulp.src('src/**/*.ts')
```
Comment (basarat) [inline gulpfile.js]: 🐛 Fixed to excluded non front-end code :rose:

File: gulpfile.js [hunk 4]
```diff
@@ -24,9 +24,7 @@ function onError(error) {
 }
 
 let projectTypings = ts.createProject('src/tsconfig.json');
-let projectCommonjs = ts.createProject('src/tsconfig.json', {
-    target: 'es6',
-});
```
Comment (basarat) [inline gulpfile.js]: I already checked that the mentioned `tsconfig.json` already has `target:es6` so you don't have to worry about me deleting this line :rose:

**Comment by basarat** [comment]
@nchanged ready for review :heart: 

 - [x] Moved key front-end modules to TypeScript
 - [x] Cleaned up gulpfile with a few minor fixes 
 - [x] Builds now fail on TypeScript errors so you / contributors can worry a bit less about potentially bad refactoring PRs (e.g. https://github.com/fuse-box/fuse-box/pull/171#issuecomment-279884006) :rose:

**Comment by nchanged** [comment]
Hi! :fire: will check it soon! Thanks!!!

**Comment by nchanged** [comment]
That's awesome! That actually fixes websocket es6 bug! (not able to minifiy)

**Pull request merged**